Testing docs (#435)
CanopyTax/single-spa.js.org: website/versioned_docs/version-5.x/testing/e2e.md

---
id: e2e
title: E2E testing
sidebar_label: E2E testing
---

:::info

As microfrontends gain widespread adoption, testing tools will catch up and the testing story will improve.

:::

> End to End (E2E) testing a single-spa [application](/docs/module-types#applications), [parcel](/docs/module-types#parcels), or [utility](/docs/module-types/#utilities) is very similar to E2E testing in other architectures. Because you are testing in the browser you can even use tools like [import-map-overrides](https://github.com/joeldenning/import-map-overrides) to run your tests in a production or production like environment with an override _before_ deloying to that environment. 

Reviewer: Let's address whether e2e testing should be done on one microfrontend at a time or not. That's the main question people have. If it's done for one microfrontend at a time, the question is how to get that set up so that it still works even with cross microfrontend imports. The standalone plugin can help in that scenario.

Also, cypress.io is the most popular e2e testing framework right now I think, so we could maybe link to it.

Author: I actually struggled with this part which is why I "skipped it". I'm torn on how to suggest it to work.

The part that's getting me confused and twisted is how to conceptualize them.

Unit tests - Jest
Integration tests - ?
End to End tests - Cypress in a real environment

I think it's arguable that you could also do integration tests with Cypress in a real environment to test the integration between modules but I don't know if that's the best way to do it.

What would be _killer_ is a JSDOM environment that SystemJS could run in and you could run Integration tests without running a browser, but that gets complicated as well.

Reviewer: I think our aim should be just to explain options, rather than prescribing which one fits best for everyone. We could talk about using import-map-overrides query params for e2e tests on a deployed environment, and also talk about using the standalone plugin + cypress as a way of testing things one at a time.

Author: This relates to why I struggled to expound on this area.

In my mind the definition of End to End tests is roughly:

> Tests that simulate what user sees and does. They should test everything in a production-like environment making actual network requests, using an actual browser, etc.

So I struggle with using `standalone` mode for this because it's not a production-like environment. I see standalone mode as a potential way to do integration tests but I worry that it's too limited to actually test many different integration points.

Perhaps my approach to this is too strict/limited and I'm not seeing the value of standalone mode.

Author: So in a vacuum I'd say something like this:

> Standalone mode is not recommended for writing E2E tests. We suggest all E2E tests run in an environment that is a close to production as possible and while standalone mode has other values it cannot deliver an experience that is close enough to production to effectively test in an end to end scenario.

But I'm not sure that's right and I wonder if the definitions I've been using to help me understand testing are wrong or don't translate well to MFEs

Reviewer: I don't think standalone mode is discouraged for e2e tests - it's an option that some are already doing and have found useful.

Author: Yup, that's my point. I have a bit of a block on standalone mode. I don't see how it's actually an end to end test in standalone mode. End to end to me implies no mocking, but standalone mode only works _because of mocking_

Reviewer: Standalone mode doesn't mock anything - it generally uses a live, deployed import map to resolve cross microfrontend imports. The main difference between standalone mode and a deployed environment is in the application registrations - on a deployed environment it registers all the applications whereas in standalone mode it only registers the one application. The benefit of standalone mode is basically just "get a root config for free that lets you see just this one microfrontend." But it doesn't require any mocking.

---
id: e2e
title: E2E testing
sidebar_label: E2E testing
---

:::info

As microfrontends gain widespread adoption, testing tools will catch up and the testing story will improve.

:::

> End to End (E2E) testing a single-spa [application](/docs/module-types#applications), [parcel](/docs/module-types#parcels), or [utility](/docs/module-types/#utilities) is very similar to E2E testing in other architectures. Because you are testing in the browser you can even use tools like [import-map-overrides](https://github.com/joeldenning/import-map-overrides) to run your tests in a production or production like environment with an override _before_ deloying to that environment. 

Reviewer: ```suggestion
End to End (E2E) testing a single-spa [application](/docs/module-types#applications), [parcel](/docs/module-types#parcels), or [utility](/docs/module-types/#utilities) is very similar to E2E testing in other architectures. Because you are testing in the browser you can even use tools like [import-map-overrides](https://github.com/joeldenning/import-map-overrides) to run your tests in a production or production like environment with an override _before_ deploying to that environment. 
```

In general we suggest only using E2E tests to test integration points between microfrontends and core functionality following principles of either the [testing pryamid](https://www.browserstack.com/guide/testing-pyramid-for-test-automation) or the [testing trophy](https://kentcdodds.com/blog/write-tests).

## Testing Options

> In single spa there are more ways to test your code in a browser using tools like [cypress](https://www.cypress.io/). Two common approachs are to test individual applications by using 'standalone' mode and testing everything together, both provide value in different ways.

Reviewer: ```suggestion
In single-spa, there are more ways to test your code in a browser using tools like [cypress](https://www.cypress.io/). Two common approaches are to test individual applications by using standalone mode and testing everything together, both provide value in different ways.
```

### "E2E" testing with "standalone" mode

> While not perfect "standalone" mode offers a way to run individual single-spa applications, it can also be used to test a single-spa application. There are several limitations to take into account though, if the microfrontend relies upon configuration or initialization happening in your single-spa `root-config` you cannot test those areas in standalone mode without mocking. Standalone mode works by creating a custom single-spa root-config on the fly that will just render the one application, so the code is the same as if it were running in production but the configuration is different.

Reviewer: The quotes around standalone seem unnecessary, and you only listed one limitation so I suggested removing the part that says there are several limitations.

```suggestion
While not perfect, standalone mode offers a way to run individual single-spa applications and can be used to test a single-spa application. If the microfrontend relies upon configuration or initialization happening in your single-spa `root-config` you cannot test those areas in standalone mode without mocking. Standalone mode works by creating a custom single-spa root-config on the fly that will just render the one application, so the code is the same as if it were running in production but the configuration is different.
```
